### DISCUSSION TITLE
LevelDB 源码分析「二、基本数据结构续」 | SF-Zhou's Blog

### DISCUSSION THREAD

**Discussion by SF-Zhou** [SF-Zhou/sf-zhou.github.io#172] [Announcements]
https://sf-zhou.github.io/leveldb/leveldb_02_data_structure.html

**Comment by SF-Zhou** [comment]
前两篇说的都是局部细节，没有过多的代码依赖，容易看懂。而真正困难的部分，是梳理清楚整体的流程和结构。这时自上而下就要容易很多，比如看看文档、架构图、流程图等，不至于在庞杂的代码细节中迷失。

**Comment by zhengdixin** [comment]
M

**Comment by SF-Zhou** [comment]
> @zhengdixin
> M

这里 M 不了吧

**Comment by WingsGo** [comment]
> @SF-Zhou
> 前两篇说的都是局部细节，没有过多的代码依赖，容易看懂。而真正困难的部分，是梳理清楚整体的流程和结构。这时自上而下就要容易很多，比如看看文档、架构图、流程图等，不至于在庞杂的代码细节中迷失。

最近看大型项目的源码，对师兄这句梳理清楚整体的流程和结构，不至于在庞杂的代码细节中迷失深有体会。

**Comment by ikiwixx** [comment]
5.(...当申请大于 1KB 或用完时则申请新的 block。....)这句话我感觉应该是，(用完时则申请新的block，如需要申请的block大于1kb，则根据申请内存大小分配，否则统一分配4kb)

**Comment by SF-Zhou** [comment]
> @ikiwixx
> 5.(...当申请大于 1KB 或用完时则申请新的 block。....)这句话我感觉应该是，(用完时则申请新的block，如需要申请的block大于1kb，则根据申请内存大小分配，否则统一分配4kb)

已更新了表述。

**Comment by ericuni** [comment]
KeyMayMatch 这个函数接口的注释是
> This method must return true if the key was in the list of keys passed to CreateFilter().
 This method may return true or false if the key was not on the list, but it should aim to return false with a high probability.

但是bloom filter 又是返回true 的时候, 元素不一定在其中, 返回false 的时候一定不在其中.
两个感觉是刚好相反的, 请问这里怎么理解?

**Comment by SF-Zhou** [comment]
> @ericuni
> KeyMayMatch 这个函数接口的注释是
> > This method must return true if the key was in the list of keys passed to CreateFilter().
>  This method may return true or false if the key was not on the list, but it should aim to return false with a high probability.
> 
> 但是bloom filter 又是返回true 的时候, 元素不一定在其中, 返回false 的时候一定不在其中.
> 两个感觉是刚好相反的, 请问这里怎么理解?
> 

英文注释和你的中文表达意思是一致的，逆否命题：
"返回 `false` 的时候一定不在其中" == "在其中一定返回 `true`"

**Comment by ericuni** [comment]
> > @ericuni
> > KeyMayMatch 这个函数接口的注释是
> > > This method must return true if the key was in the list of keys passed to CreateFilter().
> > > This method may return true or false if the key was not on the list, but it should aim to return false with a high probability.
> > 
> > 
> > 但是bloom filter 又是返回true 的时候, 元素不一定在其中, 返回false 的时候一定不在其中.
> > 两个感觉是刚好相反的, 请问这里怎么理解?
> 
> 英文注释和你的中文表达意思是一致的，逆否命题：
> "返回 `false` 的时候一定不在其中" == "在其中一定返回 `true`"

懂了, 这里好绕啊
